File: ui-server/types/models/api_response.ts
import { Episode } from './episode'
import { EpisodeSearchResult } from './episode_search_result'
import { Playback } from './playback'
import { Playlist } from './playlist'
import { Podcast } from './podcast'
import { PodcastSearchResult } from './podcast_search_result'
import { User } from './user'

export class ApiResponse {
  users: User[]
  podcasts: Podcast[]
  episodes: Episode[]
  playbacks: Playback[]
  playlists: Playlist[]
  podcastSearchResults: PodcastSearchResult[]
  episodeSearchResults: EpisodeSearchResult[]
  raw: any

  constructor(j: any) {
    const data = (j['data'] || {}) as any

    this.users = (data['users'] || []).map((o: any) => new User(o))
    this.podcasts = (data['podcasts'] || []).map((o: any) => new Podcast(o))
    this.episodes = (data['episodes'] || []).map((o: any) => new Episode(o))
    this.playbacks = (data['playbacks'] || []).map((o: any) => new Playback(o))
    this.playlists = (data['playlists'] || []).map((o: any) => new Playlist(o))
    this.podcastSearchResults = (data['podcast_search_results'] || []).map(
      (o: any) => new PodcastSearchResult(o),
    )
    this.episodeSearchResults = (data['episode_search_results'] || []).map(
      (o: any) => new EpisodeSearchResult(o),
    )
    this.raw = j['raw'] || {}
  }
}

File: model/api_response.go
package model

import "encoding/json"

type ApiResponse struct {
	Status     string            `json:"status"`
	StatusCode int               `json:"-"`
	Headers    map[string]string `json:"-"`
	Data       *ApiResponseData  `json:"data,omitempty"`
	Raw        json.RawMessage   `json:"raw,omitempty"`
	Message    string            `json:"message,omitempty"`
}

type ApiResponseData struct {
	Users               []*User             `json:"users,omitempty"`
	Podcasts            []*Podcast          `json:"podcasts,omitempty"`
	Episodes            []*Episode          `json:"episodes,omitempty"`
	Playbacks           []*Playback         `json:"playbacks,omitempty"`
	Playlists           []*Playlist         `json:"playlists,omitempty"`
	Categories          []*Category         `json:"categories,omitempty"`
	SearchSuggestions   []*SearchSuggestion `json:"search_suggestions,omitempty"`
	GlobalSearchResults *SearchResponse     `json:"global_search_results,omitempty"`
}

type SearchResponse struct {
	Podcasts []*Podcast `json:"podcasts,omitempty"`
	Episodes []*Episode `json:"episodes,omitempty"`
}

func (r *ApiResponse) ToJson() []byte {
	b, err := json.Marshal(r)
	if err != nil {
		e, _ := json.Marshal(&ApiResponse{
			Status:  "error",
			Message: "failed to marshal response",
		})
		return e
	}
	return b
}
